src/Presentation/Presenters/CreditCardNotificationPresenter.php

```diff
@@ -5,27 +5,52 @@
 namespace WMDE\Fundraising\Frontend\Presentation\Presenters;
 
 use WMDE\Fundraising\Frontend\DonationContext\UseCases\CreditCardPaymentNotification\CreditCardNotificationResponse;
-use WMDE\Fundraising\Frontend\Presentation\TwigTemplate;
 
 /**
  * @licence GNU GPL v2+
  * @author Kai Nissen < kai.nissen@wikimedia.de >
  */
 class CreditCardNotificationPresenter {
 
-	private $template;
+	private const VALUE_ASSIGNMENT = '=';
+	private const ARG_SEPARATOR = "\n";
 
-	public function __construct( TwigTemplate $template ) {
-		$this->template = $template;
+	private $returnUrl;
+
+	public function __construct( string $returnUrl ) {
+		$this->returnUrl = $returnUrl;
 	}
 
 	public function present( CreditCardNotificationResponse $response ): string {
-		return $this->template->render( [
-			'donationId' => $response->getDonationId(),
-			'accessToken' => $response->getAccessToken(),
-			'successful' => $response->isSuccessful(),
-			'errorMessage' => $response->getErrorMessage()
-		] );
+		if ( $response->isSuccessful() ) {
+			$result = [
+				'status' => 'ok',
+				'url' => $this->returnUrl . '?' . http_build_query( [
+					'id' => $response->getDonationId(),
+					'accessToken' => $response->getAccessToken()
+				] ),
+				'target' => '_top',
+				'forward' => '1',
+			];
+		} else {
+			$result = [
+				'status' => 'error',
+				'msg' => $response->getErrorMessage()
+			];
+		}
+
+		return $this->render( $result );
 	}
 
+	/**
+	 * Response format expected by 3rd party
+	 *
+	 * @see https://www.micropayment.de/help/documentation/
+	 */
+	private function render( array $result ): string {
+		array_walk( $result, function( & $value, $key ) {
```

Comment:
From a performance standpoint, `array_walk` with is in-place modification is probably faster than the more functional, side-effect free version:

    $values = array_map( 
          function( $value, $key) { return $key . self::VALUE_ASSIGNMENT . $value; },
          $result,
          array_keys( $result )
    );

Feel free to leave the code as it is, it's just me getting twitchy when seeing the `&` modifier in function parameters.

Reply:
I would have preferred a more array_map-like solution but the extra call/passing of the keys made it not worthwhile.

Reply:
![image](https://user-images.githubusercontent.com/146040/34704823-35bc3968-f4fc-11e7-9f8d-800ed9ec2604.png)

Reply:
After looking at some stuff I realized array_map can take multiple arrays, so it is not that much hassle. I gave that approach a go in https://github.com/wmde/FundraisingFrontend/pull/1176. Initial result was rather verbose till I realized the reduce could be killed.